DBZ-6417 Document how to implement a custom action (#4640)
debezium/debezium: documentation/modules/ROOT/pages/configuration/signalling.adoc

* {link-prefix}:{link-oracle-connector}#oracle-incremental-snapshots[Oracle connector incremental snapshots]
* {link-prefix}:{link-postgresql-connector}#postgresql-incremental-snapshots[PostgreSQL connector incremental snapshots]
* {link-prefix}:{link-sqlserver-connector}#sqlserver-incremental-snapshots[SQL Server connector incremental snapshots]

> // Type: procedure
> [id="debezium-signaling-custom-action"]
> === Provide Custom action
> As for xref:debezium-custom-signaling-channel[custom signal channel], you can also provide your custom actions implementing the
> following interface:

Reviewer: ```suggestion
// Type: procedure
// ModuleID: debezium-signaling-defining-a-custom-action
// Title: Defining a custom signal action
[id="debezium-signaling-custom-action"]
=== Defining a custom action

You can define custom actions for a xref:debezium-custom-signaling-channel[custom signaling channel].
Custom actions enable you to extend the {prodname} signaling framework to trigger actions that are not available in the default implementation.
You can use a custom action with multiple connectors.

To define a custom signal action, you must define the following interface:
```

[source,java,indent=0]
----
@FunctionalInterface
public interface SignalAction<P extends Partition> {

    /**
     * @param signalPayload the content of the signal
     * @return true if the signal was processed
     */
    boolean arrived(SignalPayload<P> signalPayload) throws InterruptedException;
}

----

The `io.debezium.pipeline.signal.actions.SignalAction` exposes a single method where you will receive the payload sent through the signal channel.

> After that you need to provide all your custom actions through the SPI interface `io.debezium.pipeline.signal.actions.SignalActionProvider`

Reviewer: ```suggestion
After you define a custom action, you can provide custom actions through the following `io.debezium.pipeline.signal.actions.SignalActionProvider`SPI interface.
```

[source,java,indent=0]
----
public interface SignalActionProvider {

    /**
     * Create a map of signal action where the key is the name of the action.
     *
     * @param dispatcher the event dispatcher instance
     * @param connectorConfig the connector config
     * @return a concrete action
     */

    <P extends Partition> Map<String, SignalAction<P>> createActions(EventDispatcher<P, ? extends DataCollectionId> dispatcher, CommonConnectorConfig connectorConfig);
}
----

> You need to provide your implementation to just return a map of signal action where the key of the map is the name of the action that you will use as xref:debezium-signaling-description-of-required-structure-of-a-signaling-data-collection[`type`] in the signal.

Reviewer: ```suggestion
Your implementation must return a map of the signal action.
Set the map key to the name of the action.
The key is used as the xref:debezium-signaling-description-of-required-structure-of-a-signaling-data-collection[`type`] of the signal.
```

// Type: concept
[id="debezium-signaling-custom-action-core-module-dependencies"]
=== {prodname} core module dependencies

A custom actions Java project has compile dependencies on the {prodname} core module.
You must include these compile dependencies in your project's `pom.xml` file, as shown in the following example:

[source,xml]
----
<dependency>
    <groupId>io.debezium</groupId>
    <artifactId>debezium-core</artifactId>
    <version>${version.debezium}</version> // <1>
</dependency>
----
<1> `${version.debezium}` represents the version of the {prodname} connector.

Declare your provider implementation in the `META-INF/services/io.debezium.pipeline.signal.actions.SignalActionProvider` file.

> // Type: procedure
> [id="deploying-a-debezium-custom-action"]
> === Deploying a custom action

Reviewer: ```suggestion
// Type: procedure
// ModuleID: debezium-signaling-deploying-a-custom-action
// Title: Deploying a custom signal action
[id="deploying-a-debezium-custom-action"]
=== Deploying a custom action
```

* {link-prefix}:{link-oracle-connector}#oracle-incremental-snapshots[Oracle connector incremental snapshots]
* {link-prefix}:{link-postgresql-connector}#postgresql-incremental-snapshots[PostgreSQL connector incremental snapshots]
* {link-prefix}:{link-sqlserver-connector}#sqlserver-incremental-snapshots[SQL Server connector incremental snapshots]

// Type: procedure
// ModuleID: debezium-signaling-defining-a-custom-action
// Title: Defining a custom signal action
[id="debezium-signaling-custom-action"]
=== Defining a custom action

Custom actions enable you to extend the {prodname} signaling framework to trigger actions that are not available in the default implementation.
You can use a custom action with multiple connectors.

To define a custom signal action, you must define the following interface:

[source,java,indent=0]
----
@FunctionalInterface
public interface SignalAction<P extends Partition> {

    /**
     * @param signalPayload the content of the signal
     * @return true if the signal was processed
     */
    boolean arrived(SignalPayload<P> signalPayload) throws InterruptedException;
}

----

> The `io.debezium.pipeline.signal.actions.SignalAction` exposes a single method with one parameter that is the message payloads sent through the signaling channel.

Reviewer: ```suggestion
The `io.debezium.pipeline.signal.actions.SignalAction` exposes a single method with one parameter, which represents the message payloads sent through the signaling channel.
```
I didn't notice this before. If it's correct that the parameter represents the message payloads, then we should use `which` in this sentence.

> After you define a custom action, you can provide custom actions through the following `io.debezium.pipeline.signal.actions.SignalActionProvider`SPI interface.

Reviewer: ```suggestion
After you define a custom action, you can provide custom actions through the following `io.debezium.pipeline.signal.actions.SignalActionProvider`SPI interface.
```

I have one question about this. I think that we need to differentiate more between _defining a custom action_ and _providing custom actions_ .

So, in the preceding step, you add code to implement a custom action channel (not sure whether `channel` is the appopriate term here). 
After you have done that, you must specify the actual actions that you want to trigger, yes?  Can we change the word `provide` to ``specify`, or does `provide` have special meaning in this context? Maybe something like the following:
```suggestion
After you define a custom signaling action, you can use the following SPI interface to  specify the actions that you want to trigger:  `io.debezium.pipeline.signal.actions.SignalActionProvider`SPI interface.
```

Author: Is not about triggering is actually `provide` to be available and usable later sending a signal.
